scripts/conversationalQuery.coffee

```diff
@@ -0,0 +1,130 @@
+			q = 
+				q: input,
+				searchType:'image',
+				safe:'high',
+				fields:'items(link)',
+				cx: googleCseId,
+				key: googleApiKey
+			if type == 1
+				q.fileType = 'gif'
+				q.hq = 'animated'
+			# if typeof faces is 'boolean' and faces is true
+			# 	q.imgType = 'face'
+			url = 'https://www.googleapis.com/customsearch/v1'
+			msg.http(url)
+				.query(q)
+				.get() (err,res,body) ->
+				if err
+					msg.send "ERROR: #{err}"
+					return
+				if res.statusCode isnt 200
+					msg.send "BAD HTTP REQUEST! #{res.statusCode}"
+					return
+				response = JSON.parse(body)
+				if response?.items
+					image = msg.random response.items
+					cbensureImageExtension image.link
+				else
+				msg.send "Oops. I had trouble searching '#{input}'. Try later."
+				((error) ->
+					msg.robot.logger.error error.message
+					msg.robot.logger
+					.error "(see #{error.extendedHelp})" if error.extendedHelp
+				) error for error in response.error.errors if response.error?.errors
+		else
+			# Using deprecated Google image search API
+			q = v: '1.0', rsz: '8', q: input, safe: 'active'
+			q.imgtype = 'animated' if type == 1
+			# q.imgtype = 'face' if typeof faces is 'boolean' and faces is true
+			msg.http('https://ajax.googleapis.com/ajax/services/search/images')
+				.query(q)
+				.get() (err, res, body) ->
+					if err
+						msg.send "Encountered an error :( #{err}"
+						return
+					if res.statusCode isnt 200
+						msg.send "Bad HTTP response :( #{res.statusCode}"
+						return
+					images = JSON.parse(body)
+					images = images.responseData?.results
+					if images?.length > 0
+						image = msg.random images
+						cb ensureImageExtension image.unescapedUrl
+	ensureImageExtension = (url) ->
+		ext = url.split('.').pop()
+		if /(png|jpe?g|gif)/i.test(ext)
+			url
+		else
+			"#{url}#.png"
+
+	#### IMAGE GET ####
+	getImage = (msg,imag) ->
+		queryGoogle msg,imag,0, (url) ->
+			msg.send url
+
+	#### ANIMATION GET ####
+	getAnimation = (msg,anim) ->
+		queryGoogle msg,anim,1, (url) ->
+			msg.send url
+
+
+	##---- PARSERS ----##
+
+	#### google-image IN CONVERSATION ####
+	## ASSERTION
+	robot.hear /.*woz.*(animate|animation|gif|image|picture|pic|pix|img)(?:(?: for)? me| of)?(?: a)? (.*)(?!\?)/i, (msg) ->
+		if msg.match[2].match /[^\?]$/i
+			type  = msg.match[1]
+			toGet = msg.match[2]
+			getMsg(msg,type,toGet)
+	## QUESTION
+	robot.hear /.*woz.*(?:can (?:i|you)).*(animate|animation|gif|image|picture|pic|pix|img)(?: of)?(?: a)? (.*)(?:\?)/i, (msg) ->
+		type  = msg.match[1]
+		toGet = msg.match[2]
+		getMsg(msg,type,toGet)
+
+
+
+	#### RICK ASTLEY ####
+	robot.hear /never gonna give you up/i, (msg) ->
+		rick = 1
+		msg.send "Never gonna let you down"
+		robot.hear /never gonna run around and desert you/i, (msg) ->
+			if rick == 1
+				rick = 2
+				msg.send "Never gonna make you cry"
+			robot.hear /never gonna say goodbye/i, (msg) ->
+				if rick == 2
+					rick = 3
+					msg.send "Never gonna tell a lie and hurt you!"
```

Comment:
It doesnt lol just realized that you cant save state like that. the callback is a closure so it makes new listeners every time. If you put rick on line 119 and un nested those robot.hears it would work.... I think

woz> never gonna say goodbye
woz> never gonna run around and desert you
woz> never gonna give you up
Never gonna let you down
woz> never gonna give you up
Never gonna let you down
woz> never gonna run around and desert you
Never gonna make you cry
Never gonna make you cry
woz> never gonna give you up
Never gonna let you down
woz> never gonna give you up
Never gonna let you down
woz> never gonna give you up
Never gonna let you down
woz> never gonna give you up
Never gonna let you down
woz> never gonna run around and desert you
Never gonna make you cry
Never gonna make you cry
Never gonna make you cry
Never gonna make you cry